README.md
# image-defer.js

image-defer is a simple, fast, dependency-free JavaScript library for lazily loading images on a web page. In a nutshell, images that you choose to defer are not loaded until the user can see them, while images that remain "off the page" are never loaded.

Lazily loading images has these up-sides:

* Reduced server load, server bandwidth, and client-side bandwidth, by not loading images that the user never sees
* Flatter server load, by not requesting all images at once
* Faster loading web pages
* Reduced client-side memory usage - initially at least
* Capped client-side memory usage - with `ImageDefer.options.maxLoaded`

and these down-sides:

* Requires a little extra CPU power when the user scrolls the web page
* You need to know the image dimensions in advance (use fixed sizes so that the page does not jump around as images load)
* The user may have to wait a short time for images to load when they come into view
* The user needs to have JavaScript enabled to view the images

This library was developed for, and is bundled with, the [Quru Image Server](https://github.com/quru/qis), but functions perfectly well on its own, for use with either static or dynamic images.

Unlike some other "modern" lazy loading libraries, this one does not require babel, webpack, rollup, npm, node, grunt, gulp, or any of that business.

## Live demo

You can try out image-defer on the live demo page at [https://quru.github.io/image-defer/](https://quru.github.io/image-defer/).

View the source if you want to see the plumbing, and clear your browser cache before refreshing the page if you want the images to load again remotely rather than coming from the local browser cache.

_Acknowledgements_: the placeholder and loading icons are modified from the Essential Collection pack, by _madebyoliver_ at [flaticon.com](http://flaticon.com/).

## How to use

### Adding the script

First include the image-defer script in your web page HTML:

    <script type="text/javascript" src="//your.server/path/to/image-defer.min.js" defer></script>

The script has no dependencies so you can use the `async` script tag attribute if you wish.
It does not need to run until the DOM is complete, so you can also use the `defer` script tag
attribute (unrelated to image-defer), as in the example above. See
[MDN](https://developer.mozilla.org/en/docs/Web/HTML/Element/script) for more information about
these options.

The library creates a single global object called `ImageDefer`, so it should be safe to use
alongside or to combine with any other JavaScript library that does not try to create the same.

### Defining the images

In your HTML, define images using the normal `img` tag, but have the initial `src` attribute point
to a common placeholder image, and define a new `data-defer-src` attribute to contain the proper
image URL:

    <img src="placeholder.jpg" data-defer-src="//your.server/path/to/final-image.jpg">

Any images that do not have a `data-defer-src` attribute will simply be ignored by image-defer.

The placeholder image should be the same for all your lazily loaded images, so that the browser
only loads one copy to display in all of them.

You need to use CSS, or `width` and `height` attributes, to define fixed sizes for all the lazy
images. This is so that the browser reserves the correct space for them in the page ahead of time.
Failure to do this causes 2 problems:

1) A really bad user experience, as the browser has to layout the web page again every time
   an image arrives, causing everything to jump around  
2) Image-defer, which previously recorded the locations of all the images, will not know that
   they have moved, and will start to lazy load from the wrong positions

If you don't care about (1) and you want to fix (2) then you can call `ImageDefer.reset()`
once the new page layout has completed.

For an example of CSS that defines a reusable placeholder independent of the image size,
see the source of the [demo page](docs/index.html). The placeholder is created with the
`spacer.png` file and the `defer` CSS class, and works for any image size. The `thumbnail`
CSS class defines the final image size, so you could easily define different classes for
different image sizes.

### Running

The library initialises and runs automatically as soon as the web page has loaded, during the
page's `DOMContentLoaded` event. If it works and you are happy with the defaults then there is
nothing else to do, but if not then read on.

In order to remain fast even for very large web pages, the library breaks down the web page into
horizontal bands, and then deals only with the currently visible bands. What this means in
practice is that image-defer only supports lazy loading together with vertical scrolling. This
covers most use cases. If your page scrolls horizontally, then images will be loaded regardless
of whether they are visible or horizontally off the page. This should be compatible with most
sideways-animating image slider (carousel) controls. This design is also why you might notice
horizontal groups of images loading and unloading together.

The library detects and handles window resizing and orientation changes (on mobile) automatically.
If the page layout changes dynamically in another way, you might need to tell image-defer to update
its record of those horizontal bands and what is in them. To do this, call `ImageDefer.reset()`
whenever the page layout has changed (vertically).

If you add or remove an image dynamically at runtime, you can tell image-defer to lazy load it,
or stop lazy loading it, with the `ImageDefer.addImage()` and `ImageDefer.removeImage()` functions
(see _Available functions_ below). In fact you must call `removeImage` before deleting an image
if you want it to be garbage collected, otherwise image-defer's reference to it will cause it to
be kept in memory.

If you have a very large number of images on your page and will be using the limit feature
(i.e. having more images than `ImageDefer.options.maxLoaded`), ensure that the HTTP caching
headers on your image server are set to encourage client-side caching. If the user scrolls
from the bottom of the page back up to the top, you want the unloaded images to be lazily
loaded back in again from the browser's cache rather than being re-fetched from the server. 
That said, the correct setting of your HTTP caching headers is important for every scenario.

### Runtime events

The library emits up to 3 events for each image it finds with a `data-defer-src` attribute:

* `onImageRequested` - called as soon as the URL in `data-defer-src` has been requested.
  This event is skipped if the URL loads immediately (if it was already cached by the browser).

* `onImageLoaded` - called as soon as the URL in `data-defer-src` has finished loading
  (from the remote server or from cache) and has been displayed.

* `onImageUnloaded` - called when the `ImageDefer.options.maxLoaded` limit has been exceeded,
  for images that have been "unloaded" by setting them back to the initial placeholder.

You can subscribe to these events by supplying callback functions in `ImageDefer.options`
(see _Available options_ below). Each callback requires the following prototype:

    function callback(img) {
        // img is the image DOM element being loaded or unloaded
    }

These events can be used to update your user interface with the loading state of each image.
For an example of how to do this, see the source of the [demo page](docs/index.html).

### Available options

You can define any or all of the following attributes in `ImageDefer.options`:

* `maxLoaded` (default 100) - the maximum number of images to lazy load before starting to unload images again.
  This prevents the memory use of the web page from becoming too high. Images will only be unloaded when they
  are off-screen, so if the number of visible images exceeds the limit then the limit will be breached.  
  :warning: A common browser bug currently affects the way that image-defer implements this function,
  causing the memory use of the page to continue growing in some browsers :cry: See issue #2 for more details.

* `onImageRequested` (default none) - a callback function for the _image requested_ event, see _Runtime events_ above.

* `onImageLoaded` (default none) - a callback function for the _image loaded_ event, see _Runtime events_ above.

* `onImageUnloaded` (default none) - a callback function for the _image unloaded_ event, see _Runtime events_ above.

* `scrollingStopMillis` (default 500) - the time to wait, in milliseconds, before considering that a scrolling
  event has completed. Visible images are lazy loaded when scrolling stops.

* `scrollingSkipRate` (default 0.8) - the speed of scrolling, in pixels/millisecond, above which images will
  not be lazy loaded. Reducing this value, e.g. to `0.2`, means that lazy loading will not take place while
  the user is scrolling quite slowly. Raising it, e.g. to `1.5`, means that images will still be lazy loaded
  even when the user is scrolling quickly.

You can define `ImageDefer.options` before or after including `image-defer.js` in the page, but if image-defer
loads first then your code needs to alter the existing options object rather than replacing it with a new one.

Setting options before image-defer has loaded:

    // Pre-define image-defer if it has not yet loaded
    window.ImageDefer = window.ImageDefer || {};
    
    ImageDefer.options = {
        maxLoaded: 50,
        scrollingSkipRate: 0.5
    };

Setting options after image-defer has loaded:

    ImageDefer.options.maxLoaded = 50;
    ImageDefer.options.scrollingSkipRate = 0.5;

### Available functions

Image-defer is intended to be fully automatic in the majority of cases, but there might be times
when you need to interact with it. The following functions are considered to be public:

* `ImageDefer.supportedBrowser()` - Returns a boolean for whether the browser environment appears to
  support image-defer. If this returns `false` then image-defer will not attempt to run automatically.

* `ImageDefer.reset()` - Re-scans the page for all `img` elements, and records the positions of all
  those with a `data-defer-src` attribute for lazy loading. You may need to call this if your images
  move (vertically) at some point after the page has finished loading.

* `ImageDefer.addImage(element)` - Tells image-defer to take on the lazy loading of an `img` element.
  You may need to call this if an image is dynamically added to the page. It takes no effect if the
  image does not have a `data-defer-src` attribute.

* `ImageDefer.removeImage(element, load)` - Tells image-defer to forget about an `img` element.
  This takes no effect if `addImage` was not called first. Optionally you can pass `true` for the
  second parameter to tell image-defer to eager load the image before it stops monitoring it.

* `ImageDefer.lazyLoadImages()` - Tells image-defer to load all currently visible images. You may need
  to call this if the scroll position of the page changes without having emitted the normal scroll events.

* `ImageDefer.trimImages()` - Tells image-defer to unload images, if possible, if it has loaded more
  than the `ImageDefer.options.maxLoaded` limit. Takes no effect if the limit has not been reached.

* `ImageDefer.imagesLoaded()` - Returns the number of images that are currently loading or loaded,
  i.e. loading or displaying the image specified in the `data-defer-src` attribute.

## Browser support

Based on [MDN](https://developer.mozilla.org/)'s browser support tables,
the library should in theory support these browsers (or newer):

* IE 9 or Edge
* Firefox 4
* Chrome 7
* Safari 5.1

You can use the live demo page to see if it works for a particular browser.


CHANGELOG.md
# image-defer changelog

### v0.3 Defences against bad options

Bug fixes:

* Make `ImageDefer.options` a read-only property, to prevent it from being accidentally
  overwritten (e.g. when setting options after image-defer has initialised).
  This does not prevent the individual options themselves from being changed.
* Prevent `ImageDefer.options.scrollingStopMillis` from being 0, undefined, or too low
* Do not unload images if `ImageDefer.options.maxLoaded` is 0 or undefined

### v0.2 Missing option

Bug fix:

* It was not possible to override `scrollingSkipRate` from the options.

### v0.1 Development version

This is the first usable version, and has had limited testing.


src/image-defer.js
/*!
 * image-defer.js
 * 
 * A simple dependency-free JavaScript library for implementing deferred (lazy)
 * image loading on a web page.
 * 
 * Source code:   https://github.com/quru/image-defer/
 * Date started:  24 Apr 2017
 * By:            Matt Fozard
 * Copyright:     Quru Ltd (www.quru.com)
 * Licence:
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/
 */

// TODO Support srcset, picture element?
// TODO chrome memory profiler shows allocations when scrolling - investigate

"use strict";

var ImageDefer = ImageDefer || {};

(function (options) {

    this.version = '0.2';

    // Public options - apply or set defaults (define ImageDefer.options externally to apply them here)
    this.options = {
        maxLoaded: options.maxLoaded || 100,
        onImageRequested: options.onImageRequested || null,
        onImageLoaded: options.onImageLoaded || null,
        onImageUnloaded: options.onImageUnloaded || null,
        scrollingStopMillis: options.scrollingStopMillis || 500,
        scrollingSkipRate: options.scrollingSkipRate || 0.8
    };

    // Private internal state
    var _state = {
        imagesLoaded: 0,
        images: [],
        maxBucket: 0,
        scrolling: false,
        resizing: false,
        scrollInfo: {
            last:    { pos: 0, time: 0 },
            current: { pos: 0, time: 0 }
        },
        scrollChecker: null,
        resizeChecker: null,
        lastVisibleBuckets: {
            start: -1,
            end: -1
        }
    };

    // Private image state flags
    var _imgState = {
        unloaded: 0,
        loading: 1,
        loaded: 2,
        unloading: 3
    };

    // Runs a delayed function so as to not hold up the browser
    this.requestCallBack = function(callbackFn) {
        if (window.requestAnimationFrame) requestAnimationFrame(callbackFn);
        else setTimeout(callbackFn, 20);
    }.bind(this);

    // Loads the currently visible images
    this.lazyLoadImages = function() {
        var vpPos = this.viewportPos(),
            visibleBuckets = this.getBuckets(vpPos.top, vpPos.height),
            bucket;

        // Pause loading while the viewport is resizing
        if (_state.resizing)
            return;

        if ((visibleBuckets.start !== _state.lastVisibleBuckets.start) ||
            (visibleBuckets.end !== _state.lastVisibleBuckets.end)) {

            // Load all images in the visible buckets, if they're not already loading
            // console.log('Lazily loading images for buckets ' + visibleBuckets.start + ' to ' + visibleBuckets.end);
            for (bucket = visibleBuckets.start; bucket <= visibleBuckets.end; bucket++) {
                if (_state.images[bucket]) {
                    _state.images[bucket].forEach(function(img) {
                        this.loadImage(img);
                    }.bind(this));
                }
            }

            // Don't repeat the same buckets next time around (consecutive scroll events)
            _state.lastVisibleBuckets.start = visibleBuckets.start;
            _state.lastVisibleBuckets.end = visibleBuckets.end;
        }
    }.bind(this);

    // Unloads hidden images if the max loaded limit has been reached
    this.trimImages = function() {
        if (_state.imagesLoaded > this.options.maxLoaded) {
            var vpPos = this.viewportPos(),
                visibleBuckets = this.getBuckets(vpPos.top, vpPos.height),
                unloadRequests = 0,
                unloadTotal = (_state.imagesLoaded - this.options.maxLoaded),
                self = this,
                bucket;

            var imageVisible = function(img) {
                for (var vb = visibleBuckets.start; vb <= visibleBuckets.end; vb++) {
                    if (_state.images[vb] && _state.images[vb].indexOf(img) !== -1)
                        return true;
                }
                return false;
            };
            var unloadFn = function(bucket) {
                var unloaded = 0;
                // If the bucket exists (has images in it) and is not visible
                if (_state.images[bucket] && (bucket < visibleBuckets.start || bucket > visibleBuckets.end)) {
                    _state.images[bucket].forEach(function(img) {
                        // img is in a hidden bucket, but it could span buckets, so before
                        // unloading we need to make sure that it isn't also in a visible bucket
                        if (!imageVisible(img)) {
                            if (self.unloadImage(img))
                                unloaded++;
                        }
                    });
                }
                return unloaded;
            };

            // If the user is scrolling down, unload from the top, else vice versa
            if (_state.scrollInfo.current.pos > _state.scrollInfo.last.pos) {
                // console.log('Too many images, unloading from the top');
                for (bucket = 0; (bucket <= _state.maxBucket) && (unloadRequests < unloadTotal); bucket++) {
                    unloadRequests += unloadFn(bucket);
                }
            }
            else {
                // console.log('Too many images, unloading from the bottom');
                for (bucket = _state.maxBucket; (bucket >= 0) && (unloadRequests < unloadTotal); bucket--) {
                    unloadRequests += unloadFn(bucket);
                }
            }
        }
    }.bind(this);

    // Loads the deferred-src for a single image (no effect if already loaded)
    this.loadImage = function(img) {
        if (!img._defer_state) {
            // console.log('Requesting image ' + img._defer_final_src);
            if (!img._defer_load_event) {
                img._defer_load_event = function() { this.onImageLoad(img); }.bind(this);
                img.addEventListener('load', img._defer_load_event);
            }
            img._defer_state = _imgState.loading;
            img.setAttribute('src', img._defer_final_src);
            _state.imagesLoaded++;
            // Was the image loaded immediately from browser cache?
            if (img.complete) {
                // Did the load event fire? If not (browsers vary on this), run it manually
                if (!img._defer_state)
                    this.onImageLoad(img);
            }
            else {
                // Trigger the image requested event
                if (this.options.onImageRequested)
                    this.options.onImageRequested.call(this, img);
            }
            return true;
        }
        return false;
    }.bind(this);

    // Loads the original/placeholder src for a single image (no effect if already unloaded)
    this.unloadImage = function(img) {
        if (img._defer_state && (img._defer_state !== _imgState.unloading)) {
            // console.log('Resetting image back to ' + img._defer_initial_src);
            img._defer_state = _imgState.unloading;
            img.setAttribute('src', img._defer_initial_src);
            _state.imagesLoaded--;
            // Was the image loaded immediately from browser cache?
            if (img.complete) {
                // Did the load event fire? If not (browsers vary on this), run it manually
                if (img._defer_state)
                    this.onImageLoad(img);
            }
            return true;
        }
        return false;
    }.bind(this);

    // Handler for when a requested image src is loaded
    this.onImageLoad = function(img) {
        // Firefox intermittently generates duplicate load events, 1 with complete as
        // false and 1 with complete as true. Possibly related to caching behaviour?
        if (!img.complete)
            return;
        // Even with that, we still intermittently get duplicate events, so everything
        // that follows needs to be aware that it might get called twice.

        if (this.urlsMatch(img.src, img._defer_final_src)) {
            // Image lazy load is complete
            img._defer_state = _imgState.loaded;
            if (this.options.onImageLoaded)
                this.options.onImageLoaded.call(this, img);
            // Later - check and handle the image limit
            this.requestCallBack(this.trimImages);
        }
        else if (this.urlsMatch(img.src, img._defer_initial_src)) {
            // Image unloaded is complete (it reloaded its placeholder src)
            img._defer_state = _imgState.unloaded;
            if (this.options.onImageUnloaded)
                this.options.onImageUnloaded.call(this, img);
        }
    }.bind(this);

    // Handler for scroll events
    this.onScroll = function() {
        // Discard more than 1 event per ms (also avoids div by 0 below)
        var timeNow = Date.now();
        if (_state.scrolling && (_state.scrollInfo.current.time === timeNow))
            return;

        // Track scrolling
        _state.scrolling = true;
        _state.scrollInfo.last.pos = _state.scrollInfo.current.pos;    // Avoid object creation during
        _state.scrollInfo.last.time = _state.scrollInfo.current.time;  // repeated rapid scroll events
        _state.scrollInfo.current.pos = this.viewportPos().top;
        _state.scrollInfo.current.time = timeNow;
        // Later - detect when scrolling stops
        if (!_state.scrollChecker) {
            _state.scrollChecker = setTimeout(
                this.peekScroll,
                this.options.scrollingStopMillis
            );
        }
        // Load images on the fly if scrolling is slow enough
        var scrollRate = Math.abs((_state.scrollInfo.current.pos - _state.scrollInfo.last.pos) / (_state.scrollInfo.current.time - _state.scrollInfo.last.time));
        // console.log('Scrolled to ' + _state.scrollInfo.current.pos + ', rate ' + scrollRate + ' px/ms');
        if (scrollRate < this.options.scrollingSkipRate) {
            this.requestCallBack(this.lazyLoadImages);
        }
    }.bind(this);

    // Handler for viewport resize events
    this.onResize = function() {
        _state.resizing = true;
        // When the page layout has finished changing we need to re-compute
        // all the buckets and load any newly visible images
        if (_state.resizeChecker) {
            clearTimeout(_state.resizeChecker);
        }
        _state.resizeChecker = setTimeout(function() {
            // console.log('The viewport has been resized');
            _state.resizing = false;
            _state.resizeChecker = null;
            this.reset();
        }.bind(this), 500);
    }.bind(this);

    // Detection for when scrolling has stopped
    this.peekScroll = function() {
        var timeNow = Date.now(),
            scrollInfo = _state.scrollInfo.current;

        if ((timeNow - scrollInfo.time >= this.options.scrollingStopMillis * 0.9) &&
            (this.viewportPos().top === scrollInfo.pos)) {
            // Scrolling has stopped
            _state.scrolling = false;
            _state.scrollChecker = null;
            // console.log('Scrolling stopped, images currently loaded ' + _state.imagesLoaded);
            this.requestCallBack(this.lazyLoadImages);
            if (_state.imagesLoaded > this.options.maxLoaded)
                this.requestCallBack(this.trimImages);
        }
        else {
            // Check again in a while
            _state.scrollChecker = setTimeout(
                this.peekScroll,
                this.options.scrollingStopMillis
            );
        }
    }.bind(this);

    // To speed things up, images are indexed by their vertical location
    // This returns a zero-based numeric index value (bucket number) for a given location and height
    // which can be multiple buckets if the object spans bucket boundaries and/or multiple buckets
    this.getBuckets = function(top, height) {
        var bucketSize = 200;
        return {
            start: Math.floor(top / bucketSize),
            end: Math.floor((top + height) / bucketSize)
        };
    };

    // Returns the vertical scroll position and height of the browser's visual viewport
    this.viewportPos = function() {
        return {
            top: window.pageYOffset || document.documentElement.scrollTop || 0,
            height: window.innerHeight || document.documentElement.offsetHeight || 0
        };
    };

    // Returns the vertical location and height of a DOM element
    this.elementPos = function(element) {
        var eRect = element.getBoundingClientRect();
        return {
            top: this.viewportPos().top + eRect.top,
            height: eRect.height
        };
    }.bind(this);

    // Returns whether 2 URLs are the same
    this.urlsMatch = function(url1, url2) {
        // The easy case
        if (url1 === url2)
            return true;
        // The case where one URL is relative and one is absolute
        this._aEl1 = this._aEl1 || document.createElement('a');
        this._aEl2 = this._aEl2 || document.createElement('a');
        this._aEl1.href = url1;  // Will be converted to absolute if relative
        this._aEl2.href = url2;  // "
        return this._aEl1.href === this._aEl2.href;
    }.bind(this);

    // Finds and sets up lazy loading for all images in the page, and loads the currently visible images
    // This function needs to be called whenever the page layout (vertical positions) changes
    this.reset = function() {
        // console.log('Resetting image list and state');
        _state.imagesLoaded = 0;
        _state.images = [];
        _state.maxBucket = 0;
        _state.lastVisibleBuckets.start = -1;
        _state.lastVisibleBuckets.end = -1;
        // Set up all images
        var images = document.querySelectorAll('img');
        for (var i = 0; i < images.length; i++) {
            this.addImage(images[i]);
        }
        // Load the images that are visible now
        this.requestCallBack(this.lazyLoadImages);
    }.bind(this);

    // Adds an image to be controlled by image-defer (no effect if the image has no data-defer-src attribute)
    this.addImage = function(element) {
        var deferSrc = element.getAttribute('data-defer-src'),
            initialSrc = element.getAttribute('src'),
            elPos, buckets, bucket;

        if (deferSrc) {
            // Add to the images list
            elPos = this.elementPos(element);
            buckets = this.getBuckets(elPos.top, elPos.height);
            for (bucket = buckets.start; bucket <= buckets.end; bucket++) {
                if (!_state.images[bucket]) {
                    // Initialise a new bucket
                    _state.images[bucket] = [];
                    if (bucket > _state.maxBucket)
                        _state.maxBucket = bucket;
                }
                _state.images[bucket].push(element);
            }
            // Set the deferred state of the image (only once, as reset() may be called repeatedly)
            if (element._defer_state === undefined) {
                element._defer_initial_src = initialSrc;
                element._defer_final_src = deferSrc;
                element._defer_state = this.urlsMatch(initialSrc, deferSrc) ? _imgState.loaded : _imgState.unloaded;
            }
            // Add to loaded image count (if loaded or loading)
            if ((element._defer_state === _imgState.loaded) || (element._defer_state === _imgState.loading))
                _state.imagesLoaded++;
            // console.log('Added image for ' + deferSrc + ' at ' + elPos.top + ', state ' + element._defer_state);
            return true;
        }
        return false;
    }.bind(this);

    // Removes an image from image-defer control, optionally loading it now (applying the data-defer-src attribute)
    this.removeImage = function(element, load) {
        var deferSrc = element._defer_final_src,
            currState = element._defer_state,
            elPos, elIdx, buckets, bucket;

        if (deferSrc) {
            // Remove from the images list
            elPos = this.elementPos(element);
            buckets = this.getBuckets(elPos.top, elPos.height);
            for (bucket = buckets.start; bucket <= buckets.end; bucket++) {
                if (_state.images[bucket]) {
                    elIdx = _state.images[bucket].indexOf(element);
                    if (elIdx !== -1)
                        _state.images[bucket].splice(elIdx, 1);
                }
            }
            // Remove event handlers
            if (element._defer_load_event)
                element.removeEventListener('load', element._defer_load_event);
            // Remove custom attrs
            delete element._defer_load_event;
            delete element._defer_initial_src;
            delete element._defer_final_src;
            delete element._defer_state;
            // Remove from loaded image count (if loaded or loading)
            if ((currState === _imgState.loaded) || (currState === _imgState.loading))
                _state.imagesLoaded--;
            else if (load)  // Load the final src if requested
                element.setAttribute('src', deferSrc);
            // console.log('Removed image for ' + deferSrc + ' at ' + elPos.top + ', state ' + currState);
        }
    }.bind(this);

    // Returns the number of images that are currently loaded and displayed
    this.imagesLoaded = function() {
        return _state.imagesLoaded;
    };

    // Returns whether image-defer functionality is supported by this browser environment
    this.supportedBrowser = function() {
        return !!(document.addEventListener && document.querySelectorAll &&
                  document.documentElement.getBoundingClientRect &&
                  Array.prototype.forEach);
    };

    // Initialises the library for the current web page
    this.init = function() {
        window.addEventListener('resize', this.onResize);
        window.addEventListener('orientationchange', this.onResize);
        document.addEventListener('scroll', this.onScroll);
        this.reset();
    }.bind(this);

}).call(ImageDefer, ImageDefer.options || {});


// Automatically initialise image-defer on page load
if (ImageDefer.supportedBrowser()) {
    if (document.readyState !== 'loading') ImageDefer.init();
    else document.addEventListener('DOMContentLoaded', ImageDefer.init);
}
else if (console && console.warn) {
    console.warn('image-defer: Unsupported browser');
}
